llvm/lib/Target/AIE/AIEAddressSpaceLowering.cpp

```diff
@@ -0,0 +1,113 @@
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+//===----------------------------------------------------------------------===//
+
+#include "AIEBaseSubtarget.h"
+#include "llvm/CodeGen/GlobalISel/CSEInfo.h"
+#include "llvm/CodeGen/GlobalISel/CSEMIRBuilder.h"
+#include "llvm/CodeGen/GlobalISel/GenericMachineInstrs.h"
+#include "llvm/CodeGen/GlobalISel/MachineIRBuilder.h"
+#include "llvm/CodeGen/MachineFunction.h"
+#include "llvm/CodeGen/MachineFunctionPass.h"
+#include "llvm/CodeGen/MachineInstrBuilder.h"
+#include "llvm/CodeGen/MachineModuleInfo.h"
+#include "llvm/CodeGen/TargetPassConfig.h"
+#include "llvm/InitializePasses.h"
+
+#define DEBUG_TYPE "aie-addrspace-lowering"
+
+using namespace llvm;
+
+static const char AIE_POINTER_RETYPE[] = "AIE Address Space Lowering Pass";
+
+namespace {
+
+class AIEAddressSpaceLowering : public MachineFunctionPass {
+public:
+  static char ID;
+  AIEAddressSpaceLowering() : MachineFunctionPass(ID) {}
+
+  bool runOnMachineFunction(MachineFunction &MF) override {
+    MachineRegisterInfo &MRI = MF.getRegInfo();
+    TargetPassConfig &TPC = getAnalysis<TargetPassConfig>();
+    // Enable CSE.
+    GISelCSEAnalysisWrapper &Wrapper =
+        getAnalysis<GISelCSEAnalysisWrapperPass>().getCSEWrapper();
+    auto *CSEInfo = &Wrapper.get(TPC.getCSEConfig());
+    std::unique_ptr<MachineIRBuilder> Builder =
+        CSEInfo ? std::make_unique<CSEMIRBuilder>()
+                : std::make_unique<MachineIRBuilder>();
+    Builder->setMF(MF);
+    MachineIRBuilder &MIB = *Builder;
+    // Set Observer
+    GISelObserverWrapper Observer;
+    if (CSEInfo) {
+      Observer.addObserver(CSEInfo);
+      MIB.setChangeObserver(Observer);
+    }
+
+    for (MachineBasicBlock &MBB : MF) {
+      for (MachineInstr &MI : MBB) {
+        processMI(MI, MRI, MIB, Observer);
+      }
+    }
+    return true;
+  }
+
+  void getAnalysisUsage(AnalysisUsage &AU) const override {
+    AU.addRequired<MachineModuleInfoWrapperPass>();
+    AU.addRequired<GISelCSEAnalysisWrapperPass>();
+    AU.addRequired<TargetPassConfig>();
+    AU.setPreservesAll();
+  }
+
+  StringRef getPassName() const override { return AIE_POINTER_RETYPE; }
+
+private:
+  void processMI(MachineInstr &MI, MachineRegisterInfo &MRI,
+                 MachineIRBuilder &MIB, GISelObserverWrapper &Observer) {
+
+    // Lower all pointers to pointers with default address space
+    for (llvm::MachineOperand &MO : MI.operands()) {
+      if (!MO.isReg())
+        continue;
+      Register Reg = MO.getReg();
+      if (MRI.getType(Reg).isPointer() && Reg.isVirtual()) {
+        Observer.changingInstr(MI);
+        MRI.setType(Reg, LLT::pointer(0, 20));
```

Comment:
Perhaps get the 20 from some architecture-dependent api?

Reply:
Can't we just get the size from the type we are lowering? If we run this pass, it means we know our pointers all point to aliasing/equivalent spaces. I guess we can just convert `LLT::pointer(x, y)` to `LLT::pointer(0, y)`?